src/jit/lower.cpp

```diff
@@ -571,6 +579,240 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
+    //  |  |  {  |  |     {  *  stmtExpr  void  (embedded) (IL   ???...  ???)
+    //  |  |  {  |  |     {  |  /--*  lclVar    int    V22 rat0          $380
+    //  |  |  {  |  |     {  \--*  st.lclVar int    V25 rat3
+    //  |  |  {  |  |  /--*  lclVar    int    V25 rat3
+    //  |  |  {  |  |  |  {  *  stmtExpr  void  (embedded) (IL   ???...  ???)
+    //  |  |  {  |  |  |  {  |  /--*  lclVar    int    V23 rat1
+    //  |  |  {  |  |  |  {  \--*  st.lclVar int    V26 rat4
+    //  |  |  {  |  |  +--*  lclVar    int    V26 rat4
+    //  |  |  {  |  +--*  gt_long   long
+    //  |  |  {  \--*  storeIndir long
+    //  |  +--*  lclVar    ref    V11 tmp9         u:3 (last use) $21c
+    //  |  +--*  lclVar    ref    V02 tmp0         u:3 $280
+    //  |  +--*  const     int    8 $4a
+    //  \--*  call help void   HELPER.CORINFO_HELP_ARRADDR_ST $205
+
+    GenTree* addrBase = tree->gtOp.gtOp1;
+    GenTree* dataHigh = gtLong->gtOp.gtOp2;
+    GenTree* dataLow = gtLong->gtOp.gtOp1;
+    GenTree* storeIndLow = tree;
+
+    // Rewrite storeIndLow tree to save only lower 32-bit data.
+    // 
+    //  |  |  {  |  /--*  lclVar    byref  V24 rat2   (address)
+    //  ...
+    //  |  |  {  |  +--*  lclVar    int    V25 rat3   (lower 32-bit data)
+    //  |  |  {  |  {  *  stmtExpr  void  (embedded) (IL   ???...  ???)
+    //  |  |  {  |  {  |  /--*  lclVar    int    V23 rat1
+    //  |  |  {  |  {  \--*  st.lclVar int    V26 rat4
+    //  |  |  {  \--*  storeIndir int
+    comp->fgSnipNode(currStmt, gtLong);
+    comp->fgSnipNode(currStmt, dataHigh);
+    storeIndLow->gtOp.gtOp2 = dataLow;
+    storeIndLow->gtType = TYP_INT;
+
+    // Construct storeIndHigh tree
+    //
+    // | | {  *stmtExpr  void  (embedded)(IL ? ? ? ... ? ? ? )
+    // | | { | / --*  lclVar    int    V26 rat4
+    // | | { | | / --*  lclVar    byref  V24 rat2
+    // | | { | +--*  lea(b + 4)  ref
+    // | | {  \--*  storeIndir int
+    GenTree* addrBaseHigh = new(comp, GT_LCL_VAR) GenTreeLclVar(GT_LCL_VAR, 
+        addrBase->TypeGet(), addrBase->AsLclVarCommon()->GetLclNum(), BAD_IL_OFFSET);
+    GenTree* addrHigh = new(comp, GT_LEA) GenTreeAddrMode(TYP_REF, addrBaseHigh, nullptr, 0, genTypeSize(TYP_INT));
+    GenTree* storeIndHigh = new(comp, GT_STOREIND) GenTreeStoreInd(TYP_INT, addrHigh, dataHigh);
+    storeIndHigh->gtFlags = storeIndLow->gtFlags;
+    storeIndHigh->gtFlags |= GTF_REVERSE_OPS;
+    storeIndHigh->CopyCosts(storeIndLow);
+
+    // External links of storeIndHigh tree
+    dataHigh->gtPrev = nullptr;
+    if (isEmbeddedStmt)
+    {
+        // If storeIndTree is an embedded statement, connect storeIndLow
+        // and dataHigh
+        storeIndLow->gtNext = dataHigh;
+        dataHigh->gtPrev = storeIndLow;
+    }
+    storeIndHigh->gtNext = linkEnd;
+    if (linkEnd)
+    {
+        linkEnd->gtPrev = storeIndHigh;
+    }
+
+    // Internal links of storeIndHigh tree
+    dataHigh->gtNext = addrBaseHigh;
```

Comment:
Can't we use `SimpleLinkNodeAfter` or something similar? This kind of manual linking hurts the readability of the code.